File: javascript/beginner/reverse-linked-list/reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    let prev = null;
    let current = head;
    
    while (current !== null) {
        const nextNode = current.next; // Store the next node
        current.next = prev;          // Reverse the current node's pointer
        prev = current;               // Move prev and current one step forward
        current = nextNode;
    }
    
    return prev;
};

module.exports = reverseList;

File: python/beginner/reverse-linked-list/reverse_linked_list.py
"""
Reverse Linked List

Reversing a singly linked list is a fundamental pointer manipulation problem. 
The goal is to change the next pointers of all nodes to point to the previous node in the list.

Time Complexity: O(n)
Space Complexity: O(1)
"""

from utils import ListNode


def reverse_list(head):
    """
    Reverses a singly linked list.

    Args:
      head: The head of the linked list.

    Returns:
      The new head of the reversed linked list.
    """
    prev = None
    current = head
    while current:
        next_node = current.next  # Store the next node
        current.next = prev       # Reverse the current node's pointer
        prev = current            # Move prev and current one step forward
        current = next_node
    return prev